### PR TITLE
Added "Document Signers" feature.

### PR DESCRIPTION
**Pull request by ZolotovArtem**
- Added Document signers applet
 - Added Document signers snippet
 - Updated README
 - Updated CHANGLELOG

### CODE REVIEW

File: lib/document/index.js [hunk 1]
```diff
@@ -1068,6 +1076,64 @@ class Document {
...
+   * @typedef {Object} DocumentSignersResponse
+   * @property {Array}  - emails of document signers
+   */
+
+  /**
+   * Retrieves all signers of specific document
+   * @param {DocumentSignersParams} data - document's signers list payload
+   * @param {function(err: ApiErrorResponse, res: DocumentSignersResponse)} [callback] - error first node.js callback
+   */
+  static signers ({ id, token }, callback) {
```
Comment (russellswift) [inline lib/document/index.js]: we will update this method logic after digging all requirements and details

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [v1.7.0] - Unreleased Phase 6
 
 ### Added
+- Implemented *Document Signers* feature with documentation and example in [README](https://github.com/signnow/SignNowNodeSDK/blob/master/README.md#document-signers)
```
Comment (russellswift) [inline CHANGELOG.md]: Rewrite it in new format pls

File: README.md [hunk 3]
```diff
@@ -433,6 +434,24 @@ api.document.remove({
...
 More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/remove-document.js)
 
+#### <a name="document-signers"></a>Get Document Signers
+
+```javascript
+api.document.signers({
+  token: 'your auth token',
+  id: 'document id',
+  options: {
+    freeFormInvites: false,
```
Comment (russellswift) [inline README.md]:
```suggestion
    freeFormInvites: false, // you can omit this option, it equals to false by default
```

File: README.md [hunk 4]
```diff
@@ -433,6 +434,24 @@ api.document.remove({
 
 More: [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/remove-document.js)
 
+#### <a name="document-signers"></a>Get Document Signers
+
```
Comment (russellswift) [inline README.md]: Tell users that they will get all signers of there are no options set

File: bin/document-signers.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
+
+/*
+ * to run document-signers applet from the project root folder type in your console:
+ * > node bin/document-signers <client_id> <client_secret> <username> <password> <document_id>
+ * <client_id>, <client_secret>, <username>, <password>, <document_id> - are required params
+ * options main flags:
+ *  --freeform-invites
+ *  --field-invite-status
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
 * --field-invite-status (the same as --field-invite-status=all) - will return a list of all field invite signers
 * --field-invite-status=pending,declined,fulfilled,created,skipped - will return a list of field invite signers by invite status(es). choose one(s) that you need
```

File: bin/document-signers.js [hunk 6]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
+
+/*
+ * to run document-signers applet from the project root folder type in your console:
+ * > node bin/document-signers <client_id> <client_secret> <username> <password> <document_id>
+ * <client_id>, <client_secret>, <username>, <password>, <document_id> - are required params
+ * options main flags:
+ *  --freeform-invites
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
 * --freeform-invites - will return a list of all freeform invite signers
```

File: bin/document-signers.js [hunk 7]
```diff
@@ -0,0 +1,79 @@
+#!/usr/bin/env node
+
+/*
+ * to run document-signers applet from the project root folder type in your console:
+ * > node bin/document-signers <client_id> <client_secret> <username> <password> <document_id>
+ * <client_id>, <client_secret>, <username>, <password>, <document_id> - are required params
+ * options main flags:
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
 * options:
```

File: bin/document-signers.js [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+ *  --freeform-invites
+ *  --field-invite-status
+ *  --payment-request-status
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
 * --payment-request-status (the same as --payment-request-status=all) - will return a list of all signers that payment requests are made for
 * --payment-request-status=created,pending,skipped,fulfilled - will return a list of signers that payment requests are made for by payment request status(es). choose one(s) that you need
```

File: bin/document-signers.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * additional options flags for --field-invite-status
```
Comment (russellswift) [inline bin/document-signers.js]: remove strings from 11 to 19

File: bin/document-signers.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+const flags = args.filter(arg => /^--/.test(arg));
+const params = args.filter(arg => !/^--/.test(arg));
+
+const freeForm = 'freeFormInvites';
+const fieldStatus = 'fieldInviteStatus';
+const payment = 'paymentRequestStatus';
+const allExceptFirst = arr => arr.slice(1).join('').split(',');
+
+const options = flags
+  .map(flag => flag.split('=')
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
  .map(flag => flag.split('=').filter(val => val !== ''))
```

File: bin/document-signers.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    .filter(val => val !== '')
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
```

File: bin/document-signers.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+  ).reduce((acc, arr) => {
```
Comment (russellswift) [inline bin/document-signers.js]:
```suggestion
  .reduce((acc, arr) => {
```

File: bin/document-signers.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+const args = process.argv.slice(2);
+const flags = args.filter(arg => /^--/.test(arg));
+const params = args.filter(arg => !/^--/.test(arg));
+
+const freeForm = 'freeFormInvites';
+const fieldStatus = 'fieldInviteStatus';
+const payment = 'paymentRequestStatus';
+const allExceptFirst = arr => arr.slice(1).join('').split(',');
+
+const options = flags
```
Comment (russellswift) [inline bin/document-signers.js]: I have an idea how to refactor it. I'll will suggest when you will be applying changes

File: lib/document/index.js [hunk 14]
```diff
@@ -59,6 +59,55 @@ const randString = x => {
   return s;
 };
 
+// get signers email by status
```
Comment (russellswift) [inline lib/document/index.js]: esdoc annotations pls!

File: lib/document/index.js [hunk 15]
```diff
@@ -59,6 +59,55 @@ const randString = x => {
...
+  }
+
+  if (typeof status === 'string' && status !== 'all') {
+    return getByStatus(arr, key, status);
+  }
+
+  if (Array.isArray(status)) {
+    const emailsByStatus = [];
+
+    status.forEach(value => {
```
Comment (russellswift) [inline lib/document/index.js]: why not .map?

File: lib/document/index.js [hunk 16]
```diff
@@ -1068,6 +1117,100 @@ class Document {
       .on('error', errorHandler(originalCallback))
       .end();
   }
+
+  /**
+   * Get Document Signers optional settings
+   * @typedef {Object} DocumentSignersOptions
+   * @property {boolean} [allSignatures = true] - if true receive all document signers email
```
Comment (russellswift) [inline lib/document/index.js]: is this required?

File: lib/document/index.js [hunk 17]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @property {boolean} [freeFormInvites = false] - if true receive free from invite signers email
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
   * @property {boolean} [freeFormInvites=false] - if `true` return free from invite signer emails
```

File: lib/document/index.js [hunk 18]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @property {string[]} [fieldInvitesStatus = null] - receive signers depends on fields status.
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
   * @property {?string[]} [fieldInvitesStatus=null] - return signers according to specified field invite status(es)
```

File: lib/document/index.js [hunk 19]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+     * Accepts list of statuses or a single status.
+     * Available statuses: all, pending, declined, fulfilled, created, skipped.
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
     * Acceptable statuses: all, pending, declined, fulfilled, created, skipped.
```

File: lib/document/index.js [hunk 20]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @property {string[]} [paymentRequestsStatus = null] - receive signers depends on payment status
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
   * @property {?string[]} [paymentRequestsStatus=null] - return signers according to specified to payment request status(es)
```

File: lib/document/index.js [hunk 21]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+     * Accepts list of statuses or a single status.
+     * Available statuses: all, pending, fulfilled, created, skipped.
```
Comment (russellswift) [inline lib/document/index.js]: why extra indentation?

File: lib/document/index.js [hunk 22]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @param {DocumentSignersParams} data - get document signers payload
+   * @param {function(err: ApiErrorResponse, res: DocumentSignersResponse)} [callback] - error first node.js callback
+   */
+  static signers (
+    {
+      id,
+      token,
+      options = {},
+    },
+    callback) {
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
```

File: lib/document/index.js [hunk 23]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * Retrieves all signers of specific document
+   * @param {DocumentSignersParams} data - get document signers payload
+   * @param {function(err: ApiErrorResponse, res: DocumentSignersResponse)} [callback] - error first node.js callback
+   */
+  static signers (
+    {
+      id,
+      token,
+      options = {},
+    },
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
    }, callback) {
```

File: lib/document/index.js [hunk 24]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @property {string[]}  - emails of document signers
+   */
+
+  /**
+   * Retrieves all signers of specific document
+   * @param {DocumentSignersParams} data - get document signers payload
+   * @param {function(err: ApiErrorResponse, res: DocumentSignersResponse)} [callback] - error first node.js callback
+   */
+  static signers (
+    {
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
```

File: lib/document/index.js [hunk 25]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+   * @typedef {Object} DocumentSignersResponse
+   * @property {string[]}  - emails of document signers
+   */
+
+  /**
+   * Retrieves all signers of specific document
+   * @param {DocumentSignersParams} data - get document signers payload
+   * @param {function(err: ApiErrorResponse, res: DocumentSignersResponse)} [callback] - error first node.js callback
+   */
+  static signers (
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
  static signers ({
```

File: lib/document/index.js [hunk 26]
```diff
@@ -59,6 +59,55 @@ const randString = x => {
...
 
+// get signers email by status
+const getByStatus = (arr, key, status) => {
+  return arr
+    .filter(field => field.status === status)
+    .map(val => val[key]);
+};
+
+// get signers email by key and status
+const getSignersEmail = (arr, key, status) => {
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
const getSignerEmails = (arr, key, status) => {
```

File: lib/document/index.js [hunk 27]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+        const {
+          freeFormInvites = false,
+          fieldInviteStatus = null,
+          paymentRequestStatus = null,
+        } = options;
+
+        const emails = [];
+
+        if (Object.keys(options).length === 0) {
+          const fieldInvitesEmails = getSignersEmail(field_invites, 'email') || [];
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
          const fieldInvitesEmails = getSignerEmails(field_invites, 'email') || [];
```

File: lib/document/index.js [hunk 28]
```diff
@@ -1068,6 +1117,100 @@ class Document {
...
+          const freeFormInvitesEmails = getSignersEmail(requests, 'signer_email') || [];
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
          const freeFormInvitesEmails = getSignerEmails(requests, 'signer_email') || [];
```

File: samples/snippets/documentSigners.js [hunk 29]
```diff
@@ -0,0 +1,53 @@
...
+  production: true, // if false then uses eval server
+});
+const getDocumentSigners = signnow.document.signers;
+
+const id = 'DOCUMENT_ID_GOES_HERE';
+const token = 'YOUR_ACCESS_TOKEN';
+
+/**
+ * @typedef {Object} Document signers options
+ * @property {boolean} [allSignatures = true] - if true receive all document signers email
```
Comment (russellswift) [inline samples/snippets/documentSigners.js]: same corrections as for feature annotations

File: bin/document-signers.js (hunk 9 above)
Comment (ZolotovArtem) [reply bin/document-signers.js]: ?

File: CHANGELOG.md [hunk 30]
```diff
@@ -8,7 +8,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [v1.7.0] - Unreleased Phase 6
 
 ### Added
-
+- Implemented *Document Signers* feature. See gow to apply it in [Documentation](https://signnow.github.io/SignNowNodeSDK/class/lib/document/index.js~Document.html#static-method-signers), [Short example](https://github.com/signnow/SignNowNodeSDK/blob/master/README.md#document-signers), [Full example](https://github.com/signnow/SignNowNodeSDK/blob/master/samples/snippets/documentSigners.js), [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/document-signers.js).
```
Comment (russellswift) [inline CHANGELOG.md]:
```suggestion

- Implemented *Document Signers* feature. See gow to apply it in [Documentation](https://signnow.github.io/SignNowNodeSDK/class/lib/document/index.js~Document.html#static-method-signers), [Short example](https://github.com/signnow/SignNowNodeSDK/blob/master/README.md#document-signers), [Full example](https://github.com/signnow/SignNowNodeSDK/blob/master/samples/snippets/documentSigners.js), [CLI applet](https://github.com/signnow/SignNowNodeSDK/blob/master/bin/document-signers.js).
```

File: bin/document-signers.js (hunk 9 above)
Comment (russellswift) [reply bin/document-signers.js]: 13-20 after applied changes

File: lib/document/index.js [hunk 31]
```diff
@@ -10,6 +10,8 @@ const {
   getMimeTypeByFileExtension,
   isDocumentTypeAcceptable,
   getFormDataBoundary,
+  getByDeclinedStatus,
+  getSignersEmail,
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
  getSignerEmails,
```

File: lib/common/index.js [hunk 32]
```diff
@@ -175,6 +175,55 @@ const isDocumentTypeAcceptable = fileExtension => acceptedDocumentExtensions.inc
...
 
+// get signers email by status
+const getByStatus = (arr, key, status) => {
+  return arr
+    .filter(field => field.status === status)
+    .map(val => val[key]);
+};
+
+// get signers email by key and status
+const getSignersEmail = (arr, key, status) => {
```
Comment (russellswift) [inline lib/common/index.js]:
```suggestion
const getSignerEmails = (arr, key, status) => {
```

File: lib/common/index.js [hunk 33]
```diff
@@ -186,4 +235,6 @@ module.exports = {
   isDocumentTypeAcceptable,
   isCSVFile,
   getFormDataBoundary,
+  getByDeclinedStatus,
+  getSignersEmail,
```
Comment (russellswift) [inline lib/common/index.js]:
```suggestion
  getSignerEmails,
```

File: lib/document/index.js [hunk 34]
```diff
@@ -1035,6 +1037,98 @@ class Document {
...
+        if (Object.keys(options).length === 0) {
+          const fieldInvitesEmails = getSignersEmail(field_invites, 'email') || [];
+          const freeFormInvitesEmails = getSignersEmail(requests, 'signer_email') || [];
+
+          emails.push(...fieldInvitesEmails);
+          emails.push(...freeFormInvitesEmails);
+        }
+
+        if (freeFormInvites) {
+          const freeFomInvites = getSignersEmail(requests, 'signer_email') || [];
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
          const freeFomInvites = getSignerEmails(requests, 'signer_email') || [];
```

File: lib/document/index.js [hunk 35]
```diff
@@ -1035,6 +1037,98 @@ class Document {
...
+
+          emails.push(...freeFomInvites);
+        }
+
+        if (fieldInviteStatus) {
+          const fieldInvites = getSignersEmail(field_invites, 'email', fieldInviteStatus) || [];
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
          const fieldInvites = getSignerEmails(field_invites, 'email', fieldInviteStatus) || [];
```

File: lib/document/index.js [hunk 36]
```diff
@@ -1035,6 +1037,98 @@ class Document {
...
+          if (fieldInvites.length > 0) {
+            emails.push(...fieldInvites);
+          } else {
+            emails.push(...fieldInvites);
+            emails.push(...declined);
+          }
+        }
+
+        if (paymentRequestStatus) {
+          const paymentRequests = getSignersEmail(field_invites, 'email', paymentRequestStatus) || [];
```
Comment (russellswift) [inline lib/document/index.js]:
```suggestion
          const paymentRequests = getSignerEmails(field_invites, 'email', paymentRequestStatus) || [];
```

**Pull request closed without merging**